python/ray/log.py

```diff
@@ -0,0 +1,195 @@
+import functools
+import inspect
+import logging
+import os
+import re
+import sys
+from logging.config import dictConfig
+from typing import Callable
+
+
+class ContextFilter(logging.Filter):
+    """A filter that adds ray context info to log records.
+
+    This filter adds a package name to append to the message as well as information
+    about what worker emitted the message, if applicable.
+    """
+
+    logger_regex = re.compile(r"ray(\.(?P<subpackage>\w+))?(\..*)?")
+    package_message_names = {
+        "air": "AIR",
+        "data": "Data",
+        "rllib": "RLlib",
+        "serve": "Serve",
+        "train": "Train",
+        "tune": "Tune",
+        "workflow": "Workflow",
+    }
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.global_worker_mode = None
+
+    def filter(self, record: logging.LogRecord) -> bool:
+        """Add context information to the log record.
+
+        This filter adds a package name from where the message was generated as
+        well as the worker IP address, if applicable.
+
+        Args:
+            record: Record to be filtered
+
+        Returns:
+            True if the record is to be logged, False otherwise. (This filter only
+            adds context, so records are always logged.)
+        """
+        match = self.logger_regex.search(record.name)
+        if match and match["subpackage"] in self.package_message_names:
+            record.package = f"[Ray {self.package_message_names[match['subpackage']]}]"
+        else:
+            record.package = ""
+
+        # Lazily load ray when this is called to avoid circular import
+        import ray
+
+        if hasattr(ray, "_private"):
+            if self.global_worker_mode is None:
+                self.global_worker_mode = ray._private.worker.global_worker.mode
+
+            if (
+                self.global_worker_mode
+                and self.global_worker_mode == ray._private.worker.WORKER_MODE
+            ):
+                record.node_ip_address = (
+                    ray.runtime_context.get_runtime_context().worker.node_ip_address
+                )
+
+        return True
+
+
+class AnsiStripFormatter(logging.Formatter):
+    """Formatter which strips ANSI escape codes from log messages.
+
+    These escape codes produce noise in output files and conflict with logging,
+    and so are stripped here.
+    """
+
+    strip_ansi_regex = re.compile(r"\x1b\[[0-9;]*m")
+
+    def format(self, record):
+        return re.sub(self.strip_ansi_regex, "", super().format(record))
+
+
+class PlainRayHandler(logging.StreamHandler):
+    """A plain log handler."""
+
+    def __init__(self):
+        super().__init__(stream=sys.stdout)
+
+        # Needed because the existing logging system causes workers to write to
+        # a file that is read by the driver. This will need to be removed when
```

Comment:
> This will need to be removed when
        # a proper logging system is set up in a way that lets workers pass
        # LogRecords to the driver.

Can we remove this comment? Still not sure if we will ever do this

Reply:
:+1: